File: gamer/utils/dice.py
"""Dice rolling utilities for D&D 5e."""

import random
from typing import List, Tuple


def roll_dice(num_dice: int, die_size: int, modifier: int = 0) -> Tuple[int, List[int]]:
    """
    Roll dice and return total and individual rolls.

    Args:
        num_dice: Number of dice to roll
        die_size: Size of dice (e.g., 6 for d6)
        modifier: Flat modifier to add to total

    Returns:
        Tuple of (total, list of individual rolls)
    """
    rolls = [random.randint(1, die_size) for _ in range(num_dice)]
    total = sum(rolls) + modifier
    return total, rolls


def roll(notation: str) -> Tuple[int, List[int], int]:
    """
    Roll dice using standard notation (e.g., '2d6+3', '1d20-1', '4d6').

    Args:
        notation: Dice notation string

    Returns:
        Tuple of (total, individual rolls, modifier)
    """
    notation = notation.lower().replace(' ', '')

    # Parse modifier
    modifier = 0
    if '+' in notation:
        parts = notation.split('+')
        notation = parts[0]
        modifier = int(parts[1])
    elif '-' in notation:
        parts = notation.split('-')
        notation = parts[0]
        modifier = -int(parts[1])

    # Parse dice
    if 'd' in notation:
        num_dice, die_size = notation.split('d')
        num_dice = int(num_dice) if num_dice else 1
        die_size = int(die_size)
    else:
        # Just a number, no dice
        return int(notation) + modifier, [], modifier

    total, rolls = roll_dice(num_dice, die_size, modifier)
    return total, rolls, modifier


def d4(num: int = 1, modifier: int = 0) -> int:
    """Roll d4(s) and return total."""
    total, _ = roll_dice(num, 4, modifier)
    return total


def d6(num: int = 1, modifier: int = 0) -> int:
    """Roll d6(s) and return total."""
    total, _ = roll_dice(num, 6, modifier)
    return total


def d8(num: int = 1, modifier: int = 0) -> int:
    """Roll d8(s) and return total."""
    total, _ = roll_dice(num, 8, modifier)
    return total


def d10(num: int = 1, modifier: int = 0) -> int:
    """Roll d10(s) and return total."""
    total, _ = roll_dice(num, 10, modifier)
    return total


def d12(num: int = 1, modifier: int = 0) -> int:
    """Roll d12(s) and return total."""
    total, _ = roll_dice(num, 12, modifier)
    return total


def d20(num: int = 1, modifier: int = 0) -> int:
    """Roll d20(s) and return total."""
    total, _ = roll_dice(num, 20, modifier)
    return total


def d100(num: int = 1, modifier: int = 0) -> int:
    """Roll d100(s) and return total."""
    total, _ = roll_dice(num, 100, modifier)
    return total


def roll_with_advantage() -> Tuple[int, int, int]:
    """Roll 2d20, return (higher, roll1, roll2)."""
    roll1 = random.randint(1, 20)
    roll2 = random.randint(1, 20)
    return max(roll1, roll2), roll1, roll2


def roll_with_disadvantage() -> Tuple[int, int, int]:
    """Roll 2d20, return (lower, roll1, roll2)."""
    roll1 = random.randint(1, 20)
    roll2 = random.randint(1, 20)
    return min(roll1, roll2), roll1, roll2


def roll_ability_scores() -> List[int]:
    """Roll ability scores using 4d6 drop lowest method."""
    scores = []
    for _ in range(6):
        rolls = [random.randint(1, 6) for _ in range(4)]
        rolls.sort(reverse=True)
        scores.append(sum(rolls[:3]))  # Drop lowest
    return scores


def standard_array() -> List[int]:
    """Return the standard array for ability scores."""
    return [15, 14, 13, 12, 10, 8]

File: ios/DnDTextRPG/DnDTextRPG/Utils/Dice.swift
//
//  Dice.swift
//  DnDTextRPG
//
//  Dice rolling utilities for D&D 5e
//

import Foundation

struct Dice {
    // MARK: - Basic Rolls

    /// Roll a single die
    static func roll(_ sides: Int) -> Int {
        return Int.random(in: 1...sides)
    }

    /// Roll multiple dice
    static func roll(_ count: Int, d sides: Int) -> [Int] {
        return (0..<count).map { _ in roll(sides) }
    }

    /// Roll multiple dice and return sum
    static func rollSum(_ count: Int, d sides: Int) -> Int {
        return roll(count, d: sides).reduce(0, +)
    }

    // MARK: - Common Dice

    static func d4() -> Int { roll(4) }
    static func d6() -> Int { roll(6) }
    static func d8() -> Int { roll(8) }
    static func d10() -> Int { roll(10) }
    static func d12() -> Int { roll(12) }
    static func d20() -> Int { roll(20) }
    static func d100() -> Int { roll(100) }

    // MARK: - D&D Specific Rolls

    /// Roll with advantage (roll 2d20, take higher)
    static func rollWithAdvantage() -> (result: Int, rolls: [Int]) {
        let rolls = [d20(), d20()]
        return (rolls.max()!, rolls)
    }

    /// Roll with disadvantage (roll 2d20, take lower)
    static func rollWithDisadvantage() -> (result: Int, rolls: [Int]) {
        let rolls = [d20(), d20()]
        return (rolls.min()!, rolls)
    }

    /// Roll ability scores using 4d6 drop lowest method
    static func rollAbilityScore() -> Int {
        let rolls = roll(4, d: 6).sorted(by: >)
        return rolls.prefix(3).reduce(0, +)
    }

    /// Generate a full set of ability scores
    static func rollAbilityScores() -> [Int] {
        return (0..<6).map { _ in rollAbilityScore() }.sorted(by: >)
    }

    // MARK: - Attack Rolls

    struct AttackResult {
        let roll: Int
        let total: Int
        let isCritical: Bool
        let isCriticalMiss: Bool
        let hits: Bool

        var description: String {
            if isCritical {
                return "CRITICAL HIT! (\(roll) + modifiers = \(total))"
            } else if isCriticalMiss {
                return "Critical Miss! (Natural 1)"
            } else if hits {
                return "Hit! (\(roll) + modifiers = \(total))"
            } else {
                return "Miss (\(roll) + modifiers = \(total))"
            }
        }
    }

    static func attackRoll(modifier: Int, targetAC: Int, advantage: Bool = false, disadvantage: Bool = false) -> AttackResult {
        let roll: Int
        if advantage && !disadvantage {
            roll = rollWithAdvantage().result
        } else if disadvantage && !advantage {
            roll = rollWithDisadvantage().result
        } else {
            roll = d20()
        }

        let total = roll + modifier
        let isCritical = roll == 20
        let isCriticalMiss = roll == 1

        // Natural 20 always hits, natural 1 always misses
        let hits = isCritical || (!isCriticalMiss && total >= targetAC)

        return AttackResult(
            roll: roll,
            total: total,
            isCritical: isCritical,
            isCriticalMiss: isCriticalMiss,
            hits: hits
        )
    }

    // MARK: - Saving Throws

    struct SavingThrowResult {
        let roll: Int
        let total: Int
        let success: Bool
        let isCriticalSuccess: Bool
        let isCriticalFailure: Bool
    }

    static func savingThrow(modifier: Int, dc: Int) -> SavingThrowResult {
        let roll = d20()
        let total = roll + modifier

        return SavingThrowResult(
            roll: roll,
            total: total,
            success: roll == 20 || (roll != 1 && total >= dc),
            isCriticalSuccess: roll == 20,
            isCriticalFailure: roll == 1
        )
    }

    // MARK: - Damage Rolls

    static func rollDamage(_ diceNotation: String) -> (total: Int, rolls: [Int]) {
        // Parse notation like "2d6+3" or "1d8"
        let cleaned = diceNotation.lowercased().replacingOccurrences(of: " ", with: "")

        var modifier = 0
        var diceString = cleaned

        // Extract modifier
        if let plusIndex = cleaned.firstIndex(of: "+") {
            let modStr = String(cleaned[cleaned.index(after: plusIndex)...])
            modifier = Int(modStr) ?? 0
            diceString = String(cleaned[..<plusIndex])
        } else if let minusIndex = cleaned.lastIndex(of: "-") {
            let modStr = String(cleaned[cleaned.index(after: minusIndex)...])
            modifier = -(Int(modStr) ?? 0)
            diceString = String(cleaned[..<minusIndex])
        }

        // Parse dice (e.g., "2d6")
        let parts = diceString.split(separator: "d")
        guard parts.count == 2,
              let count = Int(parts[0]),
              let sides = Int(parts[1]) else {
            return (modifier, [])
        }

        let rolls = roll(count, d: sides)
        let total = rolls.reduce(0, +) + modifier

        return (total, rolls)
    }

    /// Roll damage with critical hit (double dice)
    static func rollCriticalDamage(_ diceNotation: String) -> (total: Int, rolls: [Int]) {
        // Double the dice for critical hits
        let cleaned = diceNotation.lowercased().replacingOccurrences(of: " ", with: "")

        var modifier = 0
        var diceString = cleaned

        if let plusIndex = cleaned.firstIndex(of: "+") {
            let modStr = String(cleaned[cleaned.index(after: plusIndex)...])
            modifier = Int(modStr) ?? 0
            diceString = String(cleaned[..<plusIndex])
        } else if let minusIndex = cleaned.lastIndex(of: "-") {
            let modStr = String(cleaned[cleaned.index(after: minusIndex)...])
            modifier = -(Int(modStr) ?? 0)
            diceString = String(cleaned[..<minusIndex])
        }

        let parts = diceString.split(separator: "d")
        guard parts.count == 2,
              let count = Int(parts[0]),
              let sides = Int(parts[1]) else {
            return (modifier, [])
        }

        // Double the dice count for crits
        let rolls = roll(count * 2, d: sides)
        let total = rolls.reduce(0, +) + modifier

        return (total, rolls)
    }

    // MARK: - Death Saves

    struct DeathSaveResult {
        let roll: Int
        let isSuccess: Bool
        let isCriticalSuccess: Bool  // Natural 20 = regain 1 HP
        let isCriticalFailure: Bool  // Natural 1 = 2 failures
    }

    static func deathSave() -> DeathSaveResult {
        let roll = d20()
        return DeathSaveResult(
            roll: roll,
            isSuccess: roll >= 10,
            isCriticalSuccess: roll == 20,
            isCriticalFailure: roll == 1
        )
    }

    // MARK: - Initiative

    static func rollInitiative(dexModifier: Int) -> Int {
        return d20() + dexModifier
    }
}
